libraries/AP_AHRS/AP_AHRS_NavEKF.cpp

```diff
@@ -640,7 +640,69 @@ Vector3f AP_AHRS_NavEKF::wind_estimate(void) const
 // if we have an estimate
 bool AP_AHRS_NavEKF::airspeed_estimate(float &airspeed_ret) const
 {
-    return AP_AHRS_DCM::airspeed_estimate(get_active_airspeed_index(), airspeed_ret);
+    bool ret = false;
+    if (airspeed_sensor_enabled()) {
+        airspeed_ret = AP::airspeed()->get_airspeed();
+        return true;
+    }
+
+    if (!_flags.wind_estimation) {
+        return false;
+    }
+
+    // estimate it via nav velocity and wind estimates
+
+    // get wind estimates
+    Vector3f wind_vel;
+    switch (active_EKF_type()) {
+    case EKFType::NONE:
+        return AP_AHRS_DCM::airspeed_estimate(get_active_airspeed_index(), airspeed_ret);
+
+#if CONFIG_HAL_BOARD == HAL_BOARD_SITL
+    case EKFType::SITL:
+        if (EKF3.getWind(-1,wind_vel)) {
+             ret = true;
+        } else {
+            ret = false;
+        }
+        break;
+#endif
+
+#if HAL_NAVEKF2_AVAILABLE
+    case EKFType::TWO:
+        return AP_AHRS_DCM::airspeed_estimate(get_active_airspeed_index(), airspeed_ret);
+#endif
+
+#if HAL_NAVEKF3_AVAILABLE
+    case EKFType::THREE:
+        ret = EKF3.getWind(-1,wind_vel);
+        break;
+#endif
+
+#if HAL_EXTERNAL_AHRS_ENABLED
+    case EKFType::EXTERNAL:
+        return false;
+#endif
+    }
+
+    // estimate it via nav velocity and wind estimates
+    Vector3f nav_vel;
+    float true_airspeed;
+    if (ret && have_inertial_nav() && get_velocity_NED(nav_vel)) {
+        Vector3f true_airspeed_vec = nav_vel - wind_vel;
+        true_airspeed = true_airspeed_vec.length();
+        float gnd_speed = nav_vel.length();
+        if (_wind_max > 0) {
+            float tas_lim_lower = MAX(0.0f, (gnd_speed - _wind_max));
+            float tas_lim_upper = MAX(tas_lim_lower, (gnd_speed + _wind_max));
+            true_airspeed = constrain_float(true_airspeed, tas_lim_lower, tas_lim_upper);
+        } else {
+            true_airspeed = MAX(0.0f, true_airspeed);
+        }
+        airspeed_ret = true_airspeed / get_EAS2TAS();
+    }
```

Comment:
if ret is false here then we should fallback to the AP_AHRS_DCM::airspeed_estimate() call. Otherwise if EK3 can't provide a velocity but can return a wind then it will return true but not fill in the airspeed

Reply:
fixed

Reply:
I don't think it is quite fixed. If ret is true and have_inertial_nav() is false then it will still return true but not fill in the airspeed. The fallback at the end for ret==false needs to be outside that if() block